velox/expression/tests/PeeledEncodingTest.cpp

```diff
@@ -0,0 +1,408 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/PeeledEncoding.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::test;
+
+class PeeledEncodingTest : public testing::Test, public VectorTestBase {
+ public:
+  struct DictionaryWrap {
+    BufferPtr indices;
+    BufferPtr nulls;
+    int size;
+  };
+
+  static DictionaryWrap generateDictionaryWrap(
+      VectorFuzzer& fuzzer,
+      vector_size_t size) {
+    return {fuzzer.fuzzIndices(size, size), fuzzer.fuzzNulls(size), size};
+  }
+
+  // Utility function to wrap an arbitrary number of dictionary wrappings. Takes
+  // the 'base' vector to wrap, the number of wrappings, followed by an
+  // equivalent number of pointers to DictionaryWrap objects.
+  VectorPtr wrapInDictionaryLayers(VectorPtr base, int wrappings, ...) {
+    va_list args;
+    va_start(args, wrappings);
+    for (int i = 0; i < wrappings; i++) {
+      DictionaryWrap* wrap = va_arg(args, DictionaryWrap*);
```

Comment:
isn't the c++ more idiomatic way of doing this to use variadic template parameters and std::forward?